src/v/storage/offset_to_filepos.cc

```diff
@@ -226,6 +250,11 @@ ss::future<result<offset_to_file_pos_result>> convert_end_offset_to_file_pos(
 
                     return batch_consumer::consume_result::accept_batch;
                 }
+
+                if (internal::is_offset_in_batch(hdr, off_end)) {
+                    offset_inside_batch = true;
+                }
+
                 offset_found = true;
```

Comment:
Generally I wonder what if it makes sense to return false here if offset_found is false. If we've gotten here, and offset_found is false, shouldn't that mean we haven't found the offset?

Reply:
There are three possibilities, either the searched for offset is below segment start, inside the segment or above segment end.

For the latter two cases this code handles the search correctly, for those cases if we are here then the offset is in the segment, because we are processing a batch whose last offset is greater than the searched for offset. 

If we never end up here then the offset is not in segment because all offsets in segment are lower than the searched for offset. 

The case that is not handled correctly here is the first one, when all offsets in segment are higher than the searched for offset, in that case this falsely reports that the offset was found and returns the segment start offset. 

I think the current usages of this function do not run into the first case but this can be fixed to work in all cases.

Reply:
> if we are here then the offset is in the segment, because we are processing a batch whose last offset is greater than the searched for offset.

Let's say we have a segment that, due to compaction, only has offsets:

[0, 5, 10]

Now, what if we're searching for 7. Wouldn't we return `offset_found = true`, and then point `fo = 5`?

Reply:
In my understanding if we have a segment in the cloud with offsets [10-20] and local compacted segment have offsets [0, 7, 11, 15, 22] we should cut out the part with offsets [11, 15] but the `segment_meta` of the upload should have `base_offset = 10` and `committed_offset = 20`. At least this was the original intention.

Reply:
Would that explain the stuck reader then? the resulting segment would have [11, 15] but be advertised as [10, 20]. If we try to consume at offset 17, would we download this segment, and then not consume anything (skip all batches, and then return `stuck_reader_exception`)?